pacai/student/qlearningAgents.py

```diff
@@ -132,4 +131,5 @@ def final(self, state):
         # Did we finish training?
         if self.episodesSoFar == self.numTraining:
             # You might want to print your weights here for debugging.
-            util.raiseNotDefined()
+            # *** Your Code Here ***
+            raise NotImplementedError("ApproximateQAgent")
```

Comment:
With all of these, I don't see a need for a message in the error.
Since raising an exception will trigger a stack trace, they will already know who it is thrown by.
So just keep it simple like:
```
raise NotImplementedError()
```